### PR TITLE
Canonical Function signature creation in solidity

### PR DESCRIPTION
**Pull request by LefterisJP**
Creating the canonical function signature in solidity as requested by https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI

### CODE REVIEW

File: libsolidity/AST.cpp [hunk 1]
```diff
@@ -110,6 +110,16 @@ void FunctionDefinition::checkTypeRequirements()
 	m_body->checkTypeRequirements();
 }
 
+std::string FunctionDefinition::getCanonicalSignature()
```
Comment (chriseth) [inline libsolidity/AST.cpp]: Please don't use `std::` here.

File: libsolidity/AST.h [hunk 2]
```diff
@@ -277,6 +277,12 @@ class FunctionDefinition: public Declaration
 	/// Checks that all parameters have allowed types and calls checkTypeRequirements on the body.
 	void checkTypeRequirements();
 
+	/// Returns the canonical signature of the function
+	/// That consists of the name of the function followed by the
+	/// types of the arguments separated by commas all enclosed in parentheses
+	/// without any spaces
+	std::string getCanonicalSignature();
```
Comment (chriseth) [inline libsolidity/AST.h]: const

File: libsolidity/AST.h [hunk 3]
```diff
@@ -277,6 +277,12 @@ class FunctionDefinition: public Declaration
 	/// Checks that all parameters have allowed types and calls checkTypeRequirements on the body.
 	void checkTypeRequirements();
 
+	/// Returns the canonical signature of the function
```
Comment (chriseth) [inline libsolidity/AST.h]: `@returns` and please use `.` to end a sentence :-)

File: test/SolidityNameAndTypeResolution.cpp [hunk 4]
```diff
@@ -321,6 +323,23 @@ BOOST_AUTO_TEST_CASE(comparison_bitop_precedence)
 	BOOST_CHECK_NO_THROW(parseTextAndResolveNames(text));
 }
 
+BOOST_AUTO_TEST_CASE(function_canonical_signature)
```
Comment (chriseth) [inline test/SolidityNameAndTypeResolution.cpp]: Please don't forget to add a test with "uint", "hash" (which should be uint256 and hash256 in the signature) and "address" (which should stay as it is).

File: test/SolidityNameAndTypeResolution.cpp [hunk 5]
```diff
@@ -321,6 +323,23 @@ BOOST_AUTO_TEST_CASE(comparison_bitop_precedence)
...
+					   "  function foo(uint256 arg1, uint64 arg2, bool arg3) returns (uint256 ret) {\n"
+					   "    ret = arg1 + arg2;\n"
+					   "  }\n"
+					   "}\n";
+	BOOST_CHECK_NO_THROW(sourceUnit = parseTextAndResolveNames(text));
+	for (ASTPointer<ASTNode> const& node: sourceUnit->getNodes())
+		if (ContractDefinition* contract = dynamic_cast<ContractDefinition*>(node.get()))
+		{
+			auto functions = contract->getDefinedFunctions();
+			BOOST_ASSERT("foo(uint256,uint64,bool)" == functions[0]->getCanonicalSignature());
```
Comment (chriseth) [inline test/SolidityNameAndTypeResolution.cpp]: `BOOST_ASSERT_EQ` will generate better error messages here.

File: test/SolidityNameAndTypeResolution.cpp [hunk 6]
```diff
@@ -321,6 +323,23 @@ BOOST_AUTO_TEST_CASE(comparison_bitop_precedence)
 	BOOST_CHECK_NO_THROW(parseTextAndResolveNames(text));
 }
```
Comment (chriseth) [inline test/SolidityNameAndTypeResolution.cpp]: I think "NameAndTypeResolution" is not the right place for this test. What about ABIJSON?

File: libsolidity/Types.cpp [hunk 7]
```diff
@@ -452,6 +452,17 @@ unsigned FunctionType::getSizeOnStack() const
 	}
 }
 
+std::string FunctionType::getCanonicalSignature() const
+{
+	auto parameters = getParameterTypes();
+	std::string ret = "NAME("; //TODO: how to get function name from FunctionType
+
+	for (auto it = parameters.cbegin(); it != parameters.cend(); ++it)
+		ret += (*it)->toString() + (it + 1 == m_parameterTypes.end() ? "" : ",");
```
Comment (chriseth) [inline libsolidity/Types.cpp]: `cend`

File: libsolidity/Types.cpp [hunk 8]
```diff
@@ -452,6 +452,17 @@ unsigned FunctionType::getSizeOnStack() const
 	}
 }
 
+std::string FunctionType::getCanonicalSignature() const
+{
+	auto parameters = getParameterTypes();
+	std::string ret = "NAME("; //TODO: how to get function name from FunctionType
```
Comment (chriseth) [inline libsolidity/Types.cpp]: As discussed: The function type should not know about the name of its function, so it has to be prepended from a different place.

File: libsolidity/AST.cpp [hunk 9]
```diff
@@ -110,6 +110,16 @@ void FunctionDefinition::checkTypeRequirements()
...
+{
+	auto parameters = getParameters();
+	std::string ret = getName() + "(";
```
Comment (chriseth) [inline libsolidity/AST.cpp]: What about `return getName() + FunctionType(*this).getCanonicalSignature();`?

File: libsolidity/AST.h (hunk 3 above)
Comment (LefterisJP) [inline libsolidity/AST.h]: Okay but doxygen add '.' on its own if you don't add them yourself.

File: test/SolidityNameAndTypeResolution.cpp (hunk 6 above)
Comment (LefterisJP) [inline test/SolidityNameAndTypeResolution.cpp]: There is no real good fit anywhere. The ABIJson as the name implies only tests the generated JSON. Maybe the compiler tests would also make sense? But the reason I used name and type resolution is that this has to happen only after NameAndTypeResolution and as such thought it a good enough fit.

File: libsolidity/AST.cpp (hunk 9 above)
Comment (LefterisJP) [inline libsolidity/AST.cpp]: yeah good! I actually was just thinking about that. I wrote this function first and then noticed the existence of the FunctionType :)

File: test/SolidityNameAndTypeResolution.cpp (hunk 5 above)
Comment (LefterisJP) [inline test/SolidityNameAndTypeResolution.cpp]: There is no such function/macro. I assume you meant `BOOST_CHECK_EQUAL`

**Pull request merged**